lustre/ws.lua

```diff
@@ -0,0 +1,351 @@
+                        self:close(CloseCode.protocol())
+                    end
+                    local fm = Frame.pong(frame.payload):set_mask()
+                    self.socket:send(fm:encode())
+                elseif control_type == 'pong' then
+                    if not frame:is_final() then
+                        self:close(CloseCode.protocol())
+                    end
+                    pending_pong = false --TODO this functionality is not tested by the test framework
+                    frames_since_last_ping = 0
+                elseif control_type == 'close' then
+                    if not self._close_frame_sent then
+                        self:close(CloseCode.decode(frame.payload))
+                    end
+                    self.socket:close()
+                    return
+                end
+                goto continue
+            end
+
+            --should we close because we have been waiting to long for a ping?
+            -- we might not need to do this...it doesn't seem like it was prioritized
+            -- with a test case in autobahn
+            if pending_pong then
+                frames_since_last_ping = frames_since_last_ping + 1
+                if frames_since_last_ping > self.config._max_frames_without_pong then
+                    frames_since_last_ping = 0
+                    local err = self:close(CloseCode.policy(), "no pong after ping")
+                end
+            end
+
+
+            --handle fragmentation
+            if frame.header.opcode.sub == 'text' then
+                msg_type = "text"
+                if multiframe_message then --we expected a continuation message
+                    self:close(CloseCode.protocol(), "expected "..msg_type.."continuation frame")
+                    goto continue
+                end
+                if not frame:is_final() then
+                    multiframe_message = true
+                end
+            elseif frame.header.opcode.sub == 'binary' then
+                msg_type = "binary"
+                if multiframe_message then
+                    self:close(CloseCode.protocol(), "expected "..msg_type.."continuation frame")
+                    goto continue
+                end
+                if not frame:is_final() then
+                    multiframe_message = true
+                end
+            elseif frame.header.opcode.sub   == 'continue' and not multiframe_message then
+                self:close(CloseCode.protocol(), "unexpected continue frame")
+                goto continue
+            end
+            --aggregate payloads
+            if not frame:is_final() then --todo and we are continuing something
+                received_bytes = received_bytes + frame:payload_len()
+                --Dont build up a message payload that is bigger than max message size
+                if received_bytes <= self.config.max_message_size then
+                    table.insert(partial_frames, frame.payload)
+                end
```

Comment:
Probably want a todo here since we are truncating the total payload when it exceeds the maximum, deciding what we actually want to do might be a bigger conversation